File: src_knn_search/main/python/cratedb_knn_search.py
#
# Licensed to Crate.io GmbH ("Crate") under one or more contributor
# license agreements.  See the NOTICE file distributed with this work for
# additional information regarding copyright ownership.  Crate licenses
# this file to you under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.  You may
# obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
# License for the specific language governing permissions and limitations
# under the License.
#
# However, if you have executed another commercial license agreement
# with Crate these terms will supersede the license and you may use the
# software solely pursuant to the terms of the relevant commercial agreement.

"""
CrateDB Search CLI
==================
Interactive search against the `german_regions` table. Two modes:

  KNN (default)  : semantic search via OpenAI embeddings + KNN_MATCH on a
                   FLOAT_VECTOR column. Requires an OpenAI API key.
  Fulltext       : BM25 match() against fulltext-indexed text columns.
                   No OpenAI key needed. Enabled with --fulltext.

Connection parameters can come from CLI flags or environment variables.
Precedence: CLI flag > env var > built-in default.

Usage:
    # Semantic KNN
    python cratedb_knn_search.py --host 10.13.1.19 --user scott --password tiger

    # Fulltext BM25 (no OpenAI)
    python cratedb_knn_search.py --host 10.13.1.19 --user scott --password tiger --fulltext

    echo "wine country" | python cratedb_knn_search.py --host ... --fulltext

Output:
    Result table on stdout. All status messages on stderr.
"""

import argparse
import os
import sys

import psycopg


TABLE = "german_regions"

DEFAULTS = {
    "port": 5432,
    "user": "crate",
    "password": "",
    "database": "demo",
    "top_k": 5,
    "model": "text-embedding-3-small",
    "fulltext_columns": "tourism_info,transportation,economics,introduced_species",
}


def parse_args():
    p = argparse.ArgumentParser(
        description="Interactive search against a CrateDB table (KNN or fulltext).",
    )
    p.add_argument("--host", default=os.getenv("CRATEDB_HOST"),
                   help="CrateDB host (env: CRATEDB_HOST)")
    p.add_argument("--port", type=int,
                   default=int(os.getenv("CRATEDB_PORT", DEFAULTS["port"])),
                   help=f"CrateDB PostgreSQL port (env: CRATEDB_PORT, default {DEFAULTS['port']})")
    p.add_argument("--user", default=os.getenv("CRATEDB_USER", DEFAULTS["user"]),
                   help=f"CrateDB user (env: CRATEDB_USER, default {DEFAULTS['user']})")
    p.add_argument("--password", default=os.getenv("CRATEDB_PASSWORD", DEFAULTS["password"]),
                   help="CrateDB password (env: CRATEDB_PASSWORD)")
    p.add_argument("--database", default=os.getenv("CRATEDB_DB", DEFAULTS["database"]),
                   help=f"CrateDB database (env: CRATEDB_DB, default {DEFAULTS['database']})")
    p.add_argument("--top-k", type=int, default=DEFAULTS["top_k"],
                   help=f"Number of results to return (default {DEFAULTS['top_k']})")
    p.add_argument("--name-column", default="region_name",
                   help="Column holding the human-readable row name (default region_name)")
    p.add_argument("--fulltext", action="store_true",
                   help="Use BM25 fulltext match() instead of KNN. No OpenAI key needed.")
    p.add_argument("--fulltext-columns",
                   default=os.getenv("CRATEDB_FULLTEXT_COLUMNS", DEFAULTS["fulltext_columns"]),
                   help=("Comma-separated fulltext-indexed columns to match against "
                         "(env: CRATEDB_FULLTEXT_COLUMNS, default "
                         f"{DEFAULTS['fulltext_columns']})"))
    p.add_argument("--openai-key", default=os.getenv("OPENAI_API_KEY"),
                   help="OpenAI API key (env: OPENAI_API_KEY). Required unless --fulltext.")
    p.add_argument("--model", default=os.getenv("OPENAI_EMBED_MODEL", DEFAULTS["model"]),
                   help=f"OpenAI embedding model (env: OPENAI_EMBED_MODEL, default {DEFAULTS['model']})")

    args = p.parse_args()

    missing = []
    if not args.host:
        missing.append("--host / CRATEDB_HOST")
    if not args.fulltext and not args.openai_key:
        missing.append("--openai-key / OPENAI_API_KEY (or use --fulltext)")
    if missing:
        print("Error: missing required parameter(s): " + ", ".join(missing), file=sys.stderr)
        sys.exit(2)

    if not _is_safe_ident(args.name_column):
        print(f"Error: invalid column name {args.name_column!r}", file=sys.stderr)
        sys.exit(2)

    args.fulltext_column_list = [c.strip() for c in args.fulltext_columns.split(",") if c.strip()]
    if args.fulltext and not args.fulltext_column_list:
        print("Error: --fulltext-columns is empty", file=sys.stderr)
        sys.exit(2)
    for col in args.fulltext_column_list:
        if not _is_safe_ident(col):
            print(f"Error: invalid fulltext column name {col!r}", file=sys.stderr)
            sys.exit(2)

    return args


def _is_safe_ident(s: str) -> bool:
    if not s:
        return False
    return all(c.isalnum() or c == "_" for c in s)


def connect(args):
    return psycopg.connect(
        host=args.host,
        port=args.port,
        user=args.user,
        password=args.password,
        dbname=args.database,
        autocommit=True,
    )


def get_embedding(client, text: str, model: str) -> list:
    text = text.replace("\n", " ").strip()
    response = client.embeddings.create(input=text, model=model)
    return response.data[0].embedding


def preflight_knn(conn) -> None:
    with conn.cursor() as cur:
        cur.execute("""
            SELECT COUNT(*)
            FROM   information_schema.columns
            WHERE  table_name  = 'german_regions'
            AND    column_name = 'embedding'
        """)
        if cur.fetchone()[0] == 0:
            print("Error: table 'german_regions' has no 'embedding' column", file=sys.stderr)
            sys.exit(1)

        cur.execute("SELECT COUNT(*) FROM german_regions WHERE embedding IS NOT NULL")
        n = cur.fetchone()[0]
        if n == 0:
            print("Error: table 'german_regions' has no rows with a populated embedding", file=sys.stderr)
            sys.exit(1)
        print(f"[info] german_regions: {n} row(s) with embeddings", file=sys.stderr)


def preflight_fulltext(conn, columns: list) -> None:
    with conn.cursor() as cur:
        cur.execute("""
            SELECT column_name
            FROM   information_schema.columns
            WHERE  table_name = 'german_regions'
        """)
        present = {r[0] for r in cur.fetchall()}
    missing = [c for c in columns if c not in present]
    if missing:
        print(f"Error: table 'german_regions' is missing fulltext column(s): {missing}", file=sys.stderr)
        sys.exit(1)
    print(f"[info] fulltext match across: {', '.join(columns)}", file=sys.stderr)


def knn_search(conn, client, args, query: str):
    print(f'[info] embedding query: "{query}"', file=sys.stderr)
    vec = get_embedding(client, query, args.model)

    sql = (
        f"SELECT {args.name_column}, _score "
        f"FROM   {TABLE} "
        f"WHERE  KNN_MATCH(embedding, %s, %s) "
        f"ORDER  BY _score DESC "
        f"LIMIT  %s"
    )
    with conn.cursor() as cur:
        cur.execute(sql, (vec, args.top_k, args.top_k))
        rows = cur.fetchall()

    _print_results(rows)


def fulltext_search(conn, args, query: str):
    print(f'[info] fulltext query: "{query}"', file=sys.stderr)
    columns_sql = ", ".join(args.fulltext_column_list)

    sql = (
        f"SELECT {args.name_column}, _score "
        f"FROM   {TABLE} "
        f"WHERE  match(({columns_sql}), %s) "
        f"ORDER  BY _score DESC "
        f"LIMIT  %s"
    )
    with conn.cursor() as cur:
        cur.execute(sql, (query, args.top_k))
        rows = cur.fetchall()

    _print_results(rows)


def _print_results(rows):
    print(f"{'Region':<30}  Score")
    print(f"{'-' * 30}  -----")
    for rank, (name, score) in enumerate(rows, 1):
        print(f"#{rank}  {str(name):<28}  {score:.4f}")


def read_query() -> str:
    try:
        if sys.stdin.isatty():
            return input("Search: ").strip()
        return sys.stdin.readline().strip()
    except EOFError:
        return ""


def main():
    args = parse_args()

    query = read_query()
    if not query:
        print("[info] empty query, exiting", file=sys.stderr)
        return 0

    conn = connect(args)
    try:
        if args.fulltext:
            preflight_fulltext(conn, args.fulltext_column_list)
            fulltext_search(conn, args, query)
        else:
            from openai import OpenAI
            client = OpenAI(api_key=args.openai_key)
            preflight_knn(conn)
            knn_search(conn, client, args, query)
    finally:
        conn.close()
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: sda/src/src_knn_search/main/java/CrateDbKnnSearch.java
/*
 * Licensed to Crate.io GmbH ("Crate") under one or more contributor
 * license agreements.  See the NOTICE file distributed with this work for
 * additional information regarding copyright ownership.  Crate licenses
 * this file to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.  You may
 * obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * However, if you have executed another commercial license agreement
 * with Crate these terms will supersede the license and you may use the
 * software solely pursuant to the terms of the relevant commercial agreement.
 */

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.sql.*;
import java.util.*;

/**
 * Interactive search CLI for CrateDB's {@code german_regions} table.
 * <p>
 * Supports two modes:
 * <ul>
 *   <li><b>KNN (default)</b> — semantic search via OpenAI embeddings and
 *       CrateDB's {@code KNN_MATCH} on a {@code FLOAT_VECTOR} column.</li>
 *   <li><b>Fulltext</b> — BM25 relevance search via CrateDB's {@code MATCH}
 *       predicate. No OpenAI key needed. Enabled with {@code --fulltext}.</li>
 * </ul>
 *
 * <h2>Usage</h2>
 * <pre>
 *   # KNN (semantic) search
 *   mvn compile exec:java -Dexec.args="--host myhost --user scott --password tiger"
 *
 *   # Fulltext (BM25) search
 *   mvn compile exec:java -Dexec.args="--host myhost --user scott --password tiger --fulltext"
 * </pre>
 */
public class CrateDbKnnSearch {

    private static final String TABLE = "german_regions";
    private static final int DEFAULT_PORT = 5432;
    private static final String DEFAULT_USER = "crate";
    private static final String DEFAULT_PASSWORD = "";
    private static final String DEFAULT_DATABASE = "demo";
    private static final int DEFAULT_TOP_K = 5;
    private static final String DEFAULT_MODEL = "text-embedding-3-small";
    private static final String DEFAULT_FULLTEXT_COLUMNS =
            "tourism_info,transportation,economics,introduced_species";
    private static final String DEFAULT_NAME_COLUMN = "region_name";

    private record Config(
            String host, int port, String user, String password, String database,
            int topK, String nameColumn, boolean fulltext, List<String> fulltextColumns,
            String openaiKey, String model) {}

    public static void main(String[] args) {
        for (String arg : args) {
            if ("--help".equals(arg) || "-h".equals(arg)) {
                printUsage();
                return;
            }
        }

        Config config = parseArgs(args);

        String query = readQuery();
        if (query.isEmpty()) {
            System.err.println("[info] empty query, exiting");
            return;
        }

        try (Connection conn = connect(config)) {
            if (config.fulltext()) {
                preflightFulltext(conn, config.fulltextColumns());
                fulltextSearch(conn, config, query);
            } else {
                preflightKnn(conn);
                knnSearch(conn, config, query);
            }
        } catch (SQLException e) {
            System.err.println("Database error: " + e.getMessage());
            System.exit(1);
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            System.exit(1);
        }
    }

    private static Config parseArgs(String[] args) {
        Map<String, String> flags = new HashMap<>();
        Set<String> boolFlags = new HashSet<>();

        for (int i = 0; i < args.length; i++) {
            if ("--fulltext".equals(args[i])) {
                boolFlags.add("fulltext");
            } else if (args[i].startsWith("--") && i + 1 < args.length) {
                flags.put(args[i].substring(2), args[i + 1]);
                i++;
            }
        }

        String host = flags.getOrDefault("host", envOrDefault("CRATEDB_HOST", null));
        int port = Integer.parseInt(
                flags.getOrDefault("port", envOrDefault("CRATEDB_PORT",
                        String.valueOf(DEFAULT_PORT))));
        String user = flags.getOrDefault("user", envOrDefault("CRATEDB_USER", DEFAULT_USER));
        String password = flags.getOrDefault("password",
                envOrDefault("CRATEDB_PASSWORD", DEFAULT_PASSWORD));
        String database = flags.getOrDefault("database",
                envOrDefault("CRATEDB_DB", DEFAULT_DATABASE));
        int topK = Integer.parseInt(
                flags.getOrDefault("top-k", String.valueOf(DEFAULT_TOP_K)));
        String nameColumn = flags.getOrDefault("name-column", DEFAULT_NAME_COLUMN);
        boolean fulltext = boolFlags.contains("fulltext");
        String fulltextColumnsStr = flags.getOrDefault("fulltext-columns",
                envOrDefault("CRATEDB_FULLTEXT_COLUMNS", DEFAULT_FULLTEXT_COLUMNS));
        String openaiKey = flags.getOrDefault("openai-key",
                envOrDefault("OPENAI_API_KEY", null));
        String model = flags.getOrDefault("model",
                envOrDefault("OPENAI_EMBED_MODEL", DEFAULT_MODEL));

        List<String> missing = new ArrayList<>();
        if (host == null || host.isEmpty()) {
            missing.add("--host / CRATEDB_HOST");
        }
        if (!fulltext && (openaiKey == null || openaiKey.isEmpty())) {
            missing.add("--openai-key / OPENAI_API_KEY (or use --fulltext)");
        }
        if (!missing.isEmpty()) {
            System.err.println("Error: missing required parameter(s): "
                    + String.join(", ", missing));
            System.exit(2);
        }

        if (!isSafeIdent(nameColumn)) {
            System.err.println("Error: invalid column name '" + nameColumn + "'");
            System.exit(2);
        }

        List<String> fulltextColumnList = Arrays.stream(fulltextColumnsStr.split(","))
                .map(String::trim)
                .filter(s -> !s.isEmpty())
                .toList();

        if (fulltext && fulltextColumnList.isEmpty()) {
            System.err.println("Error: --fulltext-columns is empty");
            System.exit(2);
        }
        for (String col : fulltextColumnList) {
            if (!isSafeIdent(col)) {
                System.err.println("Error: invalid fulltext column name '" + col + "'");
                System.exit(2);
            }
        }

        return new Config(host, port, user, password, database, topK, nameColumn,
                fulltext, fulltextColumnList, openaiKey, model);
    }

    private static String envOrDefault(String envVar, String defaultValue) {
        String value = System.getenv(envVar);
        return (value != null && !value.isEmpty()) ? value : defaultValue;
    }

    private static boolean isSafeIdent(String s) {
        if (s == null || s.isEmpty()) return false;
        for (char c : s.toCharArray()) {
            if (!Character.isLetterOrDigit(c) && c != '_') return false;
        }
        return true;
    }

    private static Connection connect(Config config) throws SQLException {
        String url = String.format("jdbc:postgresql://%s:%d/%s",
                config.host(), config.port(), config.database());
        Properties props = new Properties();
        props.setProperty("user", config.user());
        props.setProperty("password", config.password());
        return DriverManager.getConnection(url, props);
    }

    private static float[] getEmbedding(String apiKey, String text, String model)
            throws Exception {
        text = text.replace("\n", " ").trim();

        JsonObject body = new JsonObject();
        body.addProperty("input", text);
        body.addProperty("model", model);

        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create("https://api.openai.com/v1/embeddings"))
                .header("Content-Type", "application/json")
                .header("Authorization", "Bearer " + apiKey)
                .POST(HttpRequest.BodyPublishers.ofString(body.toString()))
                .build();

        HttpClient client = HttpClient.newHttpClient();
        HttpResponse<String> response = client.send(request,
                HttpResponse.BodyHandlers.ofString());

        if (response.statusCode() != 200) {
            throw new RuntimeException("OpenAI API error (HTTP " + response.statusCode()
                    + "): " + response.body());
        }

        JsonObject json = JsonParser.parseString(response.body()).getAsJsonObject();
        JsonArray embeddingArray = json.getAsJsonArray("data")
                .get(0).getAsJsonObject()
                .getAsJsonArray("embedding");

        float[] embedding = new float[embeddingArray.size()];
        for (int i = 0; i < embeddingArray.size(); i++) {
            embedding[i] = embeddingArray.get(i).getAsFloat();
        }
        return embedding;
    }

    private static void preflightKnn(Connection conn) throws SQLException {
        try (Statement stmt = conn.createStatement();
             ResultSet rs = stmt.executeQuery(
                     "SELECT COUNT(*) FROM information_schema.columns "
                     + "WHERE table_name = 'german_regions' AND column_name = 'embedding'")) {
            if (rs.next() && rs.getInt(1) == 0) {
                System.err.println(
                        "Error: table 'german_regions' has no 'embedding' column");
                System.exit(1);
            }
        }

        try (Statement stmt = conn.createStatement();
             ResultSet rs = stmt.executeQuery(
                     "SELECT COUNT(*) FROM german_regions WHERE embedding IS NOT NULL")) {
            if (rs.next()) {
                int n = rs.getInt(1);
                if (n == 0) {
                    System.err.println("Error: table 'german_regions' has no rows "
                            + "with a populated embedding");
                    System.exit(1);
                }
                System.err.println("[info] german_regions: " + n
                        + " row(s) with embeddings");
            }
        }
    }

    private static void preflightFulltext(Connection conn, List<String> columns)
            throws SQLException {
        Set<String> present = new HashSet<>();
        try (Statement stmt = conn.createStatement();
             ResultSet rs = stmt.executeQuery(
                     "SELECT column_name FROM information_schema.columns "
                     + "WHERE table_name = 'german_regions'")) {
            while (rs.next()) {
                present.add(rs.getString(1));
            }
        }
        List<String> missing = columns.stream()
                .filter(c -> !present.contains(c)).toList();
        if (!missing.isEmpty()) {
            System.err.println("Error: table 'german_regions' is missing "
                    + "fulltext column(s): " + missing);
            System.exit(1);
        }
        System.err.println("[info] fulltext match across: "
                + String.join(", ", columns));
    }

    private static void knnSearch(Connection conn, Config config, String query)
            throws Exception {
        System.err.println("[info] embedding query: \"" + query + "\"");
        float[] vec = getEmbedding(config.openaiKey(), query, config.model());

        String sql = "SELECT " + config.nameColumn() + ", _score FROM " + TABLE
                + " WHERE KNN_MATCH(embedding, ?, ?) ORDER BY _score DESC LIMIT ?";

        try (PreparedStatement ps = conn.prepareStatement(sql)) {
            Float[] boxed = new Float[vec.length];
            for (int i = 0; i < vec.length; i++) boxed[i] = vec[i];
            ps.setArray(1, conn.createArrayOf("real", boxed));
            ps.setInt(2, config.topK());
            ps.setInt(3, config.topK());

            try (ResultSet rs = ps.executeQuery()) {
                printResults(rs);
            }
        }
    }

    private static void fulltextSearch(Connection conn, Config config, String query)
            throws SQLException {
        System.err.println("[info] fulltext query: \"" + query + "\"");
        String columnsSql = String.join(", ", config.fulltextColumns());

        String sql = "SELECT " + config.nameColumn() + ", _score FROM " + TABLE
                + " WHERE match((" + columnsSql + "), ?) ORDER BY _score DESC LIMIT ?";

        try (PreparedStatement ps = conn.prepareStatement(sql)) {
            ps.setString(1, query);
            ps.setInt(2, config.topK());

            try (ResultSet rs = ps.executeQuery()) {
                printResults(rs);
            }
        }
    }

    private static void printResults(ResultSet rs) throws SQLException {
        System.out.printf("%-30s  Score%n", "Region");
        System.out.printf("%-30s  -----%n", "-".repeat(30));
        int rank = 0;
        while (rs.next()) {
            rank++;
            System.out.printf("#%d  %-28s  %.4f%n",
                    rank, rs.getString(1), rs.getDouble(2));
        }
    }

    private static String readQuery() {
        try {
            if (System.console() != null) {
                String line = System.console().readLine("Search: ");
                return (line != null) ? line.trim() : "";
            }
            BufferedReader reader = new BufferedReader(
                    new InputStreamReader(System.in));
            String line = reader.readLine();
            return (line != null) ? line.trim() : "";
        } catch (Exception e) {
            return "";
        }
    }

    private static void printUsage() {
        System.err.println("Usage: CrateDbKnnSearch [options]");
        System.err.println();
        System.err.println("Options:");
        System.err.println("  --host <host>              CrateDB host (env: CRATEDB_HOST)");
        System.err.println("  --port <port>              PostgreSQL port (env: CRATEDB_PORT, default 5432)");
        System.err.println("  --user <user>              Database user (env: CRATEDB_USER, default crate)");
        System.err.println("  --password <pass>          Database password (env: CRATEDB_PASSWORD)");
        System.err.println("  --database <db>            Database name (env: CRATEDB_DB, default demo)");
        System.err.println("  --top-k <n>                Number of results (default 5)");
        System.err.println("  --name-column <col>        Column for row names (default region_name)");
        System.err.println("  --fulltext                 Use BM25 fulltext search instead of KNN");
        System.err.println("  --fulltext-columns <cols>  Comma-separated fulltext columns (env: CRATEDB_FULLTEXT_COLUMNS)");
        System.err.println("  --openai-key <key>         OpenAI API key (env: OPENAI_API_KEY, required for KNN)");
        System.err.println("  --model <model>            OpenAI embedding model (env: OPENAI_EMBED_MODEL, default text-embedding-3-small)");
        System.err.println("  -h, --help                 Show this help");
    }
}
